onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -89,6 +129,16 @@ static common::Status DeserializeTensorProto(const Env& env, const std::basic_st
 
   if (p_tensor->Location().device.Type() == OrtDevice::CPU) {
     // deserialize directly to CPU tensor
+    if (utils::HasExternalData(tensor_proto)) {
```

Comment:
Given the description says "#11127 causes the Cuda crash because when data is external, it skips the data_transfer_mgr.CopyTensor", why do we need to use ExtDataTensorProtoToTensor here if the location is CPU? 

Is the purpose of this branch to avoid a copy from the mmap'd raw data that TensorProtoToTensor does via UnpackTensor if it sees external data? If so a comment stating that would be very helpful.

Reply:
Would this be a good summary comment for this branch?

"The file containing external data for the tensor is mmap'd. If the tensor will be used on CPU we can utilize the mmap'd buffer directly by calling ExtDataTensorProtoToTensor. If we called TensorProtoToTensor it would copy the data, causing unnecessary overhead."

Without that the next developer is forced to understand all the internals of ExtDataTensorProtoToTensor and TensorProtoToTensor as well as the background mmap'ing in order to understand why this code is here.

Reply:
Hi, sorry for the late reply.
I've added an overload for `TensorProtoToTensor` that takes `OrtCallback` argument. This eliminates the need for `ExtDataTensorProtoToTensor` and other duplicated methods.

`TensorProtoToTensor` now returns early before `UnpackTensor` if data is mmap and is on CPU.

Added comment to explain this behavior.

Reply:
Previously it looks like the `tensor` parameter to TensorProtoToTensor had to be pre-allocated, but with this use case that would be unnecessary.

Given that, it may be better not to merge things into one TensorProtoToTensor implementation as the calling semantics are somewhat complicated vs having a shared file local helper for loading external data and 2 separate functions (ExtDataTensorProtoToTensor and TensorProtoToTensor) as that is a lot simpler for a user to understand. 

We should also avoid the allocation of a buffer in the Tensor instance in DeserializeTensorProto if we're doing to directly point to mmap'd external data.

Reply:
Awesome. Just a couple of minor new comments to address and I think we're good.